core/docs/features/sync-queue-design.md

```diff
@@ -0,0 +1,168 @@
+and streaming updates to the client.
+
+`SyncStreamHandler` is created for each `SyncStreams` request. It is stored in `SyncStreamHandlerRegistry` and is removed
+when `SyncStreams` request is closed.
+
+`SyncStreamHandlerRegistry` is used to look up `SyncStreamHandler` by a sync id to dispatch `ModifySync` and `CancelSync` calls.
+
+`SyncStreamHandler` sends messages to `EventBus` to subscribe or to unsubscribe from the streams.
+
+`EventBus` streams all potential updates for the requested streams to the `SyncStreamHandler` queue. When `Subscribe` message is received,
+`EventBus` requests backfill for the given stream from the `LocalSyncer` or `RemoteSyncer`. `SyncStreamHandler` can receive updates
+before relevant backfill is complete and should filter such updates out and not send them to the client.
+
+`SyncStreamHandler` and `EventBus` communicate only through message passing. `SyncStreamHandler` provides `OnHandlerUpdate` interface which
+provides relevant update methods, implementations of these methods save updates to the `SyncStreamHandler` internal queue.
+
+### `EventBus`
+
+`EventBus` is responsible for dispatching updates to the `SyncStreamHandler` queues.
+
+`LocalSyncer` and `RemoteSyncer` post updates to the `EventBus` queue through the `OnEventBusUpdate` interface.
+
+`EventBus` tracks which `SyncStreamHandler` is interested in which streams and dispatches updates to the relevant queues.
+
+When `Subscribe` message is received, `EventBus` calls `SyncerRegistry` to request backfill and start (or continue)
+update streaming. It is guaranteed that after this call either backfill or SYNC_DOWN will be posted to the `EventBus` queue
+and propagated to the `SyncStreamHandler` queues.
+
+On `Unsubscribe`, if there are no more subscribers, `EventBus` notifies `SyncerRegistry` to stop tracking this stream.
+
+### `LocalSyncer`
+
+`LocalSyncer` is responsible for streaming updates from a single local stream into the `EventBus` queue.
+
+Backfills can be requested while stream is still being loaded from the local storage. In such case `LocalSyncer`
+should save backfill requests and satisfy them once stream is loaded.
+
+### `RemoteSyncer`
+
+`RemoteSyncer` is responsible for streaming updates from a single remote stream into the `EventBus` queue.
+
+`RemoteSyncer` uses `SyncStreams` call to the remote node to request updates. It creates a new `SyncStreams` call for
+each requested stream. I.e., `RemoteSyncer` tracks a single stream and uses a single `SyncStreams` exclusively for this stream.
+
+Backfills can be requested while connection to the remote node is being established. In such case `RemoteSyncer`
+should save backfill requests and send them to the remote node once `SyncStreams` is ready and sync id is received.
+
+Lifecycle of a single `RemoteSyncer` is INIT->STREAM->SHUTDOWN. On shutdown, `SYNC_DOWN` is sent to the `EventBus`
```

Comment:
> `INIT->STREAM->SHUTDOWN`

What happens if the remote node drops off, so the transition `INIT->STREAM->REMOTE_GONE` (or is `SHUTDOWN` the state after the remote is gone and not when the local node makes a planned shutdown)?

The `RemoteSyncer` will emit the `SYNC_DOWN` update to the `EventBus` and this update is propagated to all `SyncHandlers` and eventually the clients. And once a client resubscribes to the stream a new `SyncStreamHandler` instance for that stream is started because all sync handler unsubscribed from it automatically when the stream was dropped from the lookup map in the `EventBus`?